File: windows/Protocol/Packet.cs
using System;
using System.Buffers.Binary;
using System.IO;

namespace UniversalMobileDesktop.Protocol
{
    public enum PacketType : byte
    {
        Handshake = 0x01,
        PairingRequest = 0x02,
        PairingResponse = 0x03,
        VideoFrame = 0x04,
        MouseEvent = 0x05,
        KeyEvent = 0x06,
        Clipboard = 0x07,
        FileMetadata = 0x08,
        FileChunk = 0x09,
        FileComplete = 0x0A
    }

    public class Packet
    {
        public int Length { get; set; }
        public PacketType Type { get; set; }
        public byte[] Payload { get; set; }

        public Packet(PacketType type, byte[] payload)
        {
            Type = type;
            Payload = payload ?? Array.Empty<byte>();
            Length = Payload.Length;
        }

        public byte[] Serialize()
        {
            using var ms = new MemoryStream();
            using var writer = new BinaryWriter(ms);

            Span<byte> lengthBytes = stackalloc byte[sizeof(int)];
            BinaryPrimitives.WriteInt32BigEndian(lengthBytes, Length);
            writer.Write(lengthBytes);
            writer.Write((byte)Type);
            writer.Write(Payload);
            
            return ms.ToArray();
        }

        public static Packet? Deserialize(BinaryReader reader)
        {
            try
            {
                var lengthBytes = reader.ReadBytes(sizeof(int));
                if (lengthBytes.Length != sizeof(int)) return null;
                int length = BinaryPrimitives.ReadInt32BigEndian(lengthBytes);
                if (length < 0 || length > 16 * 1024 * 1024)
                    throw new InvalidDataException($"Invalid packet length: {length}");
                byte typeByte = reader.ReadByte();
                if (!Enum.IsDefined(typeof(PacketType), typeByte))
                    throw new InvalidDataException($"Unknown packet type: 0x{typeByte:X2}");
                byte[] payload = reader.ReadBytes(length);
                if (payload.Length != length) return null;
                
                return new Packet((PacketType)typeByte, payload);
            }
            catch (EndOfStreamException)
            {
                return null;
            }
        }
    }
}

File: ios/Transport/Packet.swift
import Foundation

public enum PacketType: UInt8, Sendable {
    case handshake = 0x01
    case pairingRequest = 0x02
    case pairingResponse = 0x03
    case videoFrame = 0x04
    case mouseEvent = 0x05
    case keyEvent = 0x06
    case clipboard = 0x07
    case fileMetadata = 0x08
    case fileChunk = 0x09
    case fileComplete = 0x0A
}

public struct Packet: Equatable, Sendable {
    public static let headerSize = 5
    public static let maximumPayloadSize = 32 * 1_024 * 1_024

    public let type: PacketType
    public let payload: Data

    public init(type: PacketType, payload: Data = Data()) {
        self.type = type
        self.payload = payload
    }

    /// Encodes the Android-compatible wire format: a big-endian UInt32 length,
    /// one packet-type byte, then `length` payload bytes.
    public func encoded() throws -> Data {
        guard payload.count <= Self.maximumPayloadSize else { throw PacketError.payloadTooLarge(payload.count) }
        var length = UInt32(payload.count).bigEndian
        var data = Data(bytes: &length, count: MemoryLayout<UInt32>.size)
        data.append(type.rawValue)
        data.append(payload)
        return data
    }
}

public enum PacketError: Error, Equatable {
    case payloadTooLarge(Int)
    case unknownType(UInt8)
}

public struct PacketDecoder: Sendable {
    private var buffer = Data()

    public init() {}

    public mutating func append(_ data: Data) throws -> [Packet] {
        buffer.append(data)
        var packets: [Packet] = []

        while buffer.count >= Packet.headerSize {
            let length = buffer.prefix(4).reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
            guard length <= Packet.maximumPayloadSize else { throw PacketError.payloadTooLarge(Int(length)) }
            let packetSize = Packet.headerSize + Int(length)
            guard buffer.count >= packetSize else { break }
            let typeByte = buffer[buffer.index(buffer.startIndex, offsetBy: 4)]
            guard let type = PacketType(rawValue: typeByte) else { throw PacketError.unknownType(typeByte) }
            let payloadStart = buffer.index(buffer.startIndex, offsetBy: Packet.headerSize)
            let payloadEnd = buffer.index(payloadStart, offsetBy: Int(length))
            packets.append(Packet(type: type, payload: buffer[payloadStart..<payloadEnd]))
            buffer.removeFirst(packetSize)
        }
        return packets
    }
}
